packages/aws-cdk/lib/api/logs/logs-monitor.ts

```diff
@@ -0,0 +1,255 @@
+      error('Error occurred while monitoring logs: %s', e);
+    }
+    this.scheduleNextTick();
+  }
+
+  /**
+   * Reads all new log events from a set of CloudWatch Log Groups
+   * in parallel
+   */
+  private async readNewEvents(): Promise<Array<Array<CloudWatchLogEvent>>> {
+    const promises: Array<Promise<Array<CloudWatchLogEvent>>> = [];
+    for (const config of this.logGroups.values()) {
+      const sdk = config.sdk;
+      for (const group of config.groups) {
+        promises.push(this.readEventsFromLogGroup(sdk, group));
+      }
+    }
+    return Promise.all(promises);
+  }
+
+  /**
+   * Reads all new log events from a CloudWatch Log Group
+   * starting when the hotswap was triggered
+   *
+   * Only prints out events that have not been printed already
+   */
+  private async readEventsFromLogGroup(sdk: ISDK, logGroupName: string): Promise<CloudWatchLogEvent[]> {
+    const events: CloudWatchLogEvent[] = [];
+
+    // log events from some service are ingested faster than others
+    // so we need to track the start/end time for each log group individually
+    // to make sure that we process all events from each log group
+    const startTime = this.groupStartTime.get(logGroupName) ?? this.startTime;
+    let endTime = startTime;
+    try {
+      let nextToken: string | undefined;
+      do {
+        const response = await sdk.cloudWatchLogs().filterLogEvents({
+          logGroupName: logGroupName,
+          nextToken,
+          limit: 100,
+          startTime: startTime,
+        }).promise();
+        const eventPage = response.events ?? [];
+
+        for (const event of eventPage) {
+          if (event.message) {
+            events.push({
+              message: event.message,
+              logGroup: logGroupName,
+              timestamp: event.timestamp ? new Date(event.timestamp) : new Date(),
+            });
+
+            if (event.timestamp && endTime < event.timestamp) {
+              endTime = event.timestamp;
+            }
+          }
+        }
+        nextToken = response.nextToken;
+      } while (nextToken);
+    } catch (e) {
+      // with Lambda functions the CloudWatch is not created
+      // until something is logged, so just keep polling until
+      // there is somthing to find
+      if (e.code === 'ResourceNotFoundException') {
+        return [];
+      }
+      throw e;
+    }
+    this.groupStartTime.set(logGroupName, endTime+1);
+    return events;
+  }
+}
+
+interface PrinterProps {}
+
+export interface IEventPrinter {
+  print(event: CloudWatchLogEvent): void;
+}
+
+abstract class EventPrinterBase {
+  constructor(protected readonly props?: PrinterProps) {
+  }
+
+  public abstract print(event: CloudWatchLogEvent): void;
+}
+
+/**
+ * a CloudWatchLogs event printer
+ */
+export class EventPrinter extends EventPrinterBase {
+  constructor(props?: PrinterProps) {
+    super(props);
+  }
+
+  public print(event: CloudWatchLogEvent): void {
+    print(util.format('[%s] %s %s',
```

Comment:
I think this may be due to Lambda adding a newline character to the message. In my gif I have an example of an ECS service where the logs are printed without the empty line. I think we might be able to just trim any newlines from the message.

Reply:
Yes please!

Reply:
Done.